src/cdomains/intDomain.ml

```diff
@@ -919,22 +932,535 @@ struct
+
+  let bit f ik (i1, i2) =
+    match (interval_to_int i1), (interval_to_int i2) with
+    | Some x, Some y -> (try norm_interval ik (Interval.of_int ik (f ik x y)) with Division_by_zero -> top_of ik)
+    | _ -> top_of ik
+
+  let bit1 f ik i1 =
+    match interval_to_int i1 with
+    | Some x -> of_int ik (f ik x)
+    | _ -> top_of ik
+
+  let bitcomp f ik (i1, i2) = 
+    match (interval_to_int i1, interval_to_int i2) with 
+    | Some x, Some y -> (try norm_interval ik (Interval.of_int ik (f ik x y)) with Division_by_zero | Invalid_argument _ -> top_of ik)
+    | _, _ -> (set_overflow_flag ~cast:false ~underflow:true ~overflow:true ik;  top_of ik)
+
+  let bitand ik x y = 
+    let interval_bitand = bit (fun _ik -> Ints_t.bitand) ik in
+    binary_op x y interval_bitand
+
+  let bitor ik x y = 
+    let interval_bitor = bit (fun _ik -> Ints_t.bitor) ik in
+    binary_op x y interval_bitor
+
+  let bitxor ik x y = 
+    let interval_bitxor = bit (fun _ik -> Ints_t.bitxor) ik in
+    binary_op x y interval_bitxor
+
+  let bitnot ik x = 
+    let interval_bitnot = bit1 (fun _ik -> Ints_t.bitnot) ik in
+    unary_op x interval_bitnot
+
+  let shift_left ik x y = 
+    let interval_shiftleft = bitcomp (fun _ik x y -> Ints_t.shift_left x (Ints_t.to_int y)) ik in
+    binary_op x y interval_shiftleft
+
+  let shift_right ik x y = 
+    let interval_shiftright = bitcomp (fun _ik x y -> Ints_t.shift_right x (Ints_t.to_int y)) ik in
+    binary_op x y interval_shiftright
+
+  let lognot ik x = 
+    let interval_lognot = log1 (fun _ik -> not) ik in
+    unary_op x interval_lognot
+
+  let logand ik x y = 
+    let interval_logand = log (&&) ik in
+    binary_op x y interval_logand
+
+  let logor ik x y = 
+    let interval_logor = log (||) ik in
+    binary_op x y interval_logor
+
+  let add ?no_ov ik x y =
+    let interval_add ((x1, x2), (y1, y2)) = norm_interval ik @@ Some (Ints_t.add x1 y1, Ints_t.add x2 y2) in
+    binary_op x y interval_add
+
+  let neg ?no_ov ik x = 
+    let neg_interval ((x, y)) = norm_interval ik @@ Some (Ints_t.neg y, Ints_t.neg x) in
+    unary_op x neg_interval
+
+  let sub ?no_ov ik x y = 
+    let interval_sub ((x1, x2), (y1, y2)) = norm_interval ik @@ Some (Ints_t.sub x1 y2, Ints_t.sub x2 y1) in
+    binary_op x y interval_sub
+
+  let mul ?no_ov (ik: ikind) (x: t) (y: t) : t = 
+    let interval_mul ((x1, x2), (y1, y2)) = 
+      let x1y1 = (Ints_t.mul x1 y1) in let x1y2 = (Ints_t.mul x1 y2) in
+      let x2y1 = (Ints_t.mul x2 y1) in let x2y2 = (Ints_t.mul x2 y2) in
+      norm_interval ik @@ Some ((Ints_t.min (Ints_t.min x1y1 x1y2) (Ints_t.min x2y1 x2y2)), (Ints_t.max (Ints_t.max x1y1 x1y2) (Ints_t.max x2y1 x2y2)))
+    in
+    binary_op x y interval_mul
+
+  let div ?no_ov ik x y = 
+    let rec interval_div ((x1, x2), (y1, y2)) = begin
+      let is_zero v = Ints_t.compare v Ints_t.zero = 0 in
+      match y1, y2 with
+      | l, u when is_zero l && is_zero u -> top_of ik (* TODO warn about undefined behavior *)
```

Comment:
What's up with this TODO? The warnings should indeed be generated here, as this domain may be more precise than intervals (same logic as for the overflows).

Reply:
I believe we discussed this. Same TODO exists in IntervalFunctior. Let me know if I forgot something.